modules/Emails/EmailFromValidator.php

```diff
@@ -0,0 +1,470 @@
+                $ok = false;
+            }
+            
+            if (isset($emailObj->FromName) && $name !== $emailObj->FromName) {
+                $this->addError(self::ERR_FIELD_FROM_ADDR_NAME_INVALID_EMAIL_PART_TO_FIELD_FROMNAME);
+                $ok = false;
+            }
+            
+            if (isset($emailObj->from_name) && $name !== $emailObj->from_name) {
+                $this->addError(self::ERR_FIELD_FROM_ADDR_NAME_INVALID_EMAIL_PART_TO_FIELD_FROM_NAME);
+                $ok = false;
+            }
+            
+            if ($ok) {
+                $valid = true;
+            }
+        }
+        return $valid;
+    }
+
+
+    /**
+     * validate field 'From' - should be a valid email address
+     *
+     * @return array
+     */
+    protected function validateFrom()
+    {
+        $email = $this->getEmail();
+        if (!isset($email->From)) {
+            $this->addError(self::ERR_FIELD_FROM_IS_NOT_SET);
+        } else {
+            if (!$email->From) {
+                $this->addError(self::ERR_FIELD_FROM_IS_EMPTY);
+            } else {
+                if (!$this->isValidFromAddress($email->From)) {
+                    $this->addError(self::ERR_FIELD_FROM_IS_INVALID);
+                }
+            }
+        }
+        return $this->getErrors();
+    }
+    
+    /**
+     * validate field 'from_addr' - should be a valid email address
+     *
+     * @return array
+     */
+    protected function validateFromAddr()
+    {
+        $email = $this->getEmail();
+        if (!isset($email->from_addr)) {
+            $this->addError(self::ERR_FIELD_FROM_ADDR_IS_NOT_SET);
+        } else {
+            if (!$email->from_addr) {
+                $this->addError(self::ERR_FIELD_FROM_ADDR_IS_EMPTY);
+            } else {
+                if (!$this->isValidFromAddress($email->from_addr)) {
+                    $this->addError(self::ERR_FIELD_FROM_ADDR_IS_INVALID);
+                }
+            }
+        }
+        return $this->getErrors();
+    }
+    
+    /**
+     * validate field 'FromName' - should be a valid name string
+     *
+     * @return array
+     */
+    protected function validateFromName()
+    {
+        $email = $this->getEmail();
+        if (!isset($email->FromName)) {
+            $this->addError(self::ERR_FIELD_FROMNAME_IS_NOT_SET);
+        } else {
+            if (!$email->FromName) {
+                $this->addError(self::ERR_FIELD_FROMNAME_IS_EMPTY);
+            } else {
+                if (!$this->isValidFromName($email->FromName)) {
+                    $this->addError(self::ERR_FIELD_FROMNAME_IS_INVALID);
+                }
+            }
+        }
+        return $this->getErrors();
+    }
+    
+    /**
+     * validate field 'from_name' - should be a valid name string
+     *
+     * @return array
+     */
+    protected function validateFrom_Name()
+    {
+        $email = $this->getEmail();
```

Comment:
$email already declared

Reply:
where?

Reply:
Line 89, will get picked up as a code quality issue.

Reply:
Any tool should really pick this up, reference: https://wiki.sei.cmu.edu/confluence/display/java/DCL51-J.+Do+not+shadow+or+obscure+identifiers+in+subscopes

Reply:
java?!

Reply:
Yes, makes no difference for PHP but still included in all the tools. I can create configs to ignore this.